### PR TITLE
test(IDX): let most system-tests depend on the mainnet NNS in PRs

### PR DESCRIPTION
**Pull request by basvandijk**
What
====
This duplicates most system-tests which depend on the NNS by:
* Having the original system-test with name `<name>` now depend on the mainnet NNS instead of the HEAD NNS.
* Having a new system-test with the name `<name>_head_nns` which depends on the HEAD NNS.
 
`<name>_head_nns`  is additionally tagged with `system_test_nightly` such that it runs nightly and not on PRs.

The above is done for all system-tests which depend on the NNS except for the ones under `rs/tests/nns` and `rs/tests/nns/sns` because they test the NNS instead of the replica.

Why
===
The idea is that for most system-tests which test the replica it's more realistic to test against the mainnet NNS since that version would be active when the replica would be released. It's also useful to reduce CI times since changes to the NNS now don't trigger as many system-tests (although to achieve this fully more work like https://github.com/dfinity/ic/pull/628 is needed).

However it's still useful to see if the HEAD replica works against the HEAD NNS (see [this comment](https://gitlab.com/dfinity-lab/public/ic/-/merge_requests/20322#note_1989750018) in an earlier attempt for an example) which is why this macro introduces the `<name>_head_nns` variant which only runs daily if not overridden.

How
===

This is follow up work from https://github.com/dfinity/ic/pull/675.

It extends the `system_test(...)` bazel macro with a `test_driver_target` argument which specifies the name of the target of the test driver. It defaults to `None` which means that the macro declares its own test driver as a rust_binary from `<name>.rs`. The macro then also returns a structure specifying the `test_driver_target`.

This is then used in the new bazel macro `system_test_nns(name, ...)` which declares two system-tests: one named `<name>` depending on the mainnet NNS (as specified by [mainnet-canisters.bzl](https://github.com/dfinity/ic/blob/master/mainnet-canisters.bzl)) and one `head_nns` variant which depends on the head NNS. The latter then also gets the additional `system_test_nightly` tag which can be overridden if desired.

Tests
===

Additional testing:

* Schedule Hourly: https://github.com/dfinity/ic/actions/runs/10231299140
* Schedule Daily: https://github.com/dfinity/ic/actions/runs/10231299567
* Release Testing: https://github.com/dfinity/ic/actions/runs/10231300143

### CODE REVIEW

**Review by Bownairo — APPROVED** [review]

File: rs/tests/Cargo.toml [hunk 1]
```diff
@@ -304,12 +304,12 @@ name = "test-driver-e2e-scenarios"
...
-name = "ic-systest-boundary-node-integration"
-path = "boundary_nodes/boundary_node_integration_test.rs"
+name = "ic-systest-boundary-node-intgrtn"
+path = "boundary_nodes/boundary_node_intgrtn_test.rs"
 
 [[bin]]
-name = "ic-systest-boundary-node-integration-on-playnet"
-path = "boundary_nodes/boundary_node_integration_on_playnet_test.rs"
+name = "ic-systest-boundary-node-intgrtn-on-playnet"
+path = "boundary_nodes/boundary_node_intgrtn_on_playnet_test.rs"
```
Comment (basvandijk) [inline rs/tests/Cargo.toml:313]: Note to reviewers: the name `boundary_node_integration_on_playnet_test` has been shortened to work around a `DomainLabelTooLong` error thrown by Route53.

**Review by DSharifi — APPROVED** [review]

**Review by r-birkner — APPROVED** [review]

File: rs/tests/system_tests.bzl [hunk 2]
```diff
@@ -263,6 +268,50 @@ def system_test(
         flaky = flaky,
     )
 
+def system_test_nns(name, extra_head_nns_tags = ["system_test_nightly"], **kwargs):
+    """Declares a system-test that uses the mainnet NNS and a variant that use the HEAD NNS.
+
+    Declares two system-tests:
+
+    * One with the given name which uses the NNS from mainnet as specified by testnet/mainnet_revisions.json
```
Comment (eichhorl) [inline rs/tests/system_tests.bzl:276]: Doesn't `testnet/mainnet_revisions.json` specify the current NNS subnet _replica_ version? As I understand the git revision of mainnet NNS _canisters_ might be different

**Review by eichhorl — APPROVED** [review]

File: rs/tests/system_tests.bzl (hunk 2 above)
Comment (basvandijk) [reply rs/tests/system_tests.bzl]: Thanks for spotting that! Fixed.

File: rs/tests/system_tests.bzl [hunk 3]
```diff
@@ -129,6 +130,7 @@ def system_test(
 
     Args:
       name: base name to use for the binary and test rules.
+      variant: string to identify the variant of this test.
```
Comment (nmattia) [inline rs/tests/system_tests.bzl:133]: what's a variant? what are the possible values? if it's only used in naming, why not simply use `name_<variant>` as a `name` when calling the `system_test` and reduce complexity?

File: rs/tests/system_tests.bzl [hunk 4]
```diff
@@ -159,17 +161,18 @@ def system_test(
     """
 
     # Names are used as part of domain names; thus, limit their length
-    if len(name) > 50:
+    if len(name) > 50 or len(variant) > 25:
```
Comment (nmattia) [inline rs/tests/system_tests.bzl:164]: Can you split this into two checks with corresponding error messages, or update the message?

File: rs/tests/system_tests.bzl [hunk 5]
```diff
@@ -159,17 +161,18 @@ def system_test(
...
-        srcs = [name + ".rs"],
-        **kwargs
-    )
+    if variant == "":
+        rust_binary(
+            name = bin_name,
+            testonly = True,
+            srcs = [name + ".rs"],
+            **kwargs
+        )
```
Comment (nmattia) [inline rs/tests/system_tests.bzl:175]: What are the semantics of `variant`?

File: rs/tests/system_tests.bzl (hunk 4 above)
Comment (basvandijk) [reply rs/tests/system_tests.bzl:164]: Actually the most accurate check is:
```python
if len(name + variant_suffix) + 14 > 63:
```
but that's only relevant for tests that [create_dns_records](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/tests/driver/src/driver/farm.rs?L264). So I think it's best if we just drop this check.

File: rs/tests/system_tests.bzl (hunk 5 above)
Comment (basvandijk) [reply rs/tests/system_tests.bzl:181]: I added some docs.

File: rs/tests/system_tests.bzl [hunk 6]
```diff
@@ -263,6 +274,50 @@ def system_test(
...
+    )
+
+    original_tags = kwargs.pop("tags", [])
+    system_test(
+        name,
+        variant = "head_nns",
+        env = env | NNS_CANISTER_ENV,
+        runtime_deps = runtime_deps + NNS_CANISTER_RUNTIME_DEPS,
+        tags = [tag for tag in original_tags if tag not in extra_head_nns_tags] +
+               extra_head_nns_tags if extra_head_nns_tags != None else original_tags,
```
Comment (basvandijk) [inline rs/tests/system_tests.bzl:317]: `if extra_head_nns_tags != None` is unnecessary. If folks want to `original_tags` they can just set `extra_head_nns_tags` to `[]`.

**Review by nmattia — APPROVED** [review]
LGTM, thanks!

**Pull request merged**